### DISCUSSION TITLE
support for arm64 or work around

### DISCUSSION THREAD

**Discussion by harshameghadri** [biocore-ntnu/pyranges#346] [Ideas]
Hi,

Firstly, thank you for creating the package and the effort it requires to maintain it. 

I have recently switched from intel mac to a m1 mac with arm64. As I see now that there is no native package for pyranges/ncls for arm64. is there a work around so I can try it out on m1?

I don't have access to an intel or x86_64 system for the project I am working on.

---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
Cell In [19], line 1
----> 1 import pyranges as pr

File /opt/homebrew/lib/python3.9/site-packages/pyranges/__init__.py:17
     15 from pyranges.methods.concat import concat
     16 from pyranges.multioverlap import count_overlaps
---> 17 from pyranges.pyranges_main import PyRanges
     18 from pyranges.readers import read_bam, read_bed, read_bigwig, read_gff3, read_gtf  # NOQA: F401
     20 __version__ = pkg_resources.get_distribution("pyranges").version

File /opt/homebrew/lib/python3.9/site-packages/pyranges/pyranges_main.py:8
      5 from natsort import natsorted  # type: ignore
      7 import pyranges as pr
----> 8 from pyranges.methods.intersection import _intersection, _overlap
      9 from pyranges.multithreaded import (
     10     _extend,
     11     _extend_grp,
   (...)
     16     pyrange_apply_single,
     17 )
     18 from pyranges.tostring2 import tostring

File /opt/homebrew/lib/python3.9/site-packages/pyranges/methods/intersection.py:3
      1 import numpy as np
      2 import pandas as pd
----> 3 from ncls import NCLS  # type: ignore
      6 def _intersection(scdf, ocdf, **kwargs):
      7     how = kwargs["how"]

File ~/miniforge3/envs/torch-gpu/lib/python3.9/site-packages/ncls/__init__.py:1
----> 1 from ncls.src.ncls import NCLS64
      2 from ncls.src.ncls32 import NCLS32
      4 import numpy as np

ImportError: dlopen(/Users/sriharshameghadri/miniforge3/envs/torch-gpu/lib/python3.9/site-packages/ncls/src/ncls.cpython-39-darwin.so, 0x0002): tried: '/Users/sriharshameghadri/miniforge3/envs/torch-gpu/lib/python3.9/site-packages/ncls/src/ncls.cpython-39-darwin.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64')), '/System/Volumes/Preboot/Cryptexes/OS/Users/sriharshameghadri/miniforge3/envs/torch-gpu/lib/python3.9/site-packages/ncls/src/ncls.cpython-39-darwin.so' (no such file), '/Users/sriharshameghadri/miniforge3/envs/torch-gpu/lib/python3.9/site-packages/ncls/src/ncls.cpython-39-darwin.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64'))

**Comment by marco-mariotti** [comment]
Hi,
I routinely work with pyranges on my M2 mac.
Do you have Rosetta2 installed? If not, install it. Normally the OS would
ask you automatically to install it upon necessity, but probably this does
not happen when the need arises through command line interactions.

If the problem persists, try to install conda (miniconda) or mamba, and
install pyranges from the bioconda channel. This is how it worked for me at
the first attempt.



On Fri, Jul 21, 2023 at 11:09 AM Sri Harsha Meghadri <
***@***.***> wrote:

> Hi,
>
> Firstly, thank you for creating the package and the effort it requires to
> maintain it.
>
> I have recently switched from intel mac to a m1 mac with arm64. As I see
> now that there is no native package for pyranges/ncls for arm64. is there a
> work around so I can try it out on m1?
>
> I don't have access to an intel or x86_64 system for the project I am
> working on.
> ------------------------------
>
> ImportError Traceback (most recent call last)
> Cell In [19], line 1
> ----> 1 import pyranges as pr
>
> File /opt/homebrew/lib/python3.9/site-packages/pyranges/*init*.py:17
> 15 from pyranges.methods.concat import concat
> 16 from pyranges.multioverlap import count_overlaps
> ---> 17 from pyranges.pyranges_main import PyRanges
> 18 from pyranges.readers import read_bam, read_bed, read_bigwig,
> read_gff3, read_gtf # NOQA: F401
> 20 *version* = pkg_resources.get_distribution("pyranges").version
>
> File /opt/homebrew/lib/python3.9/site-packages/pyranges/pyranges_main.py:8
> 5 from natsort import natsorted # type: ignore
> 7 import pyranges as pr
> ----> 8 from pyranges.methods.intersection import _intersection, _overlap
> 9 from pyranges.multithreaded import (
> 10 _extend,
> 11 _extend_grp,
> (...)
> 16 pyrange_apply_single,
> 17 )
> 18 from pyranges.tostring2 import tostring
>
> File
> /opt/homebrew/lib/python3.9/site-packages/pyranges/methods/intersection.py:3
> 1 import numpy as np
> 2 import pandas as pd
> ----> 3 from ncls import NCLS # type: ignore
> 6 def _intersection(scdf, ocdf, **kwargs):
> 7 how = kwargs["how"]
>
> File ~/miniforge3/envs/torch-gpu/lib/python3.9/site-packages/ncls/*init*
> .py:1
> ----> 1 from ncls.src.ncls import NCLS64
> 2 from ncls.src.ncls32 import NCLS32
> 4 import numpy as np
>
> ImportError:
> dlopen(/Users/sriharshameghadri/miniforge3/envs/torch-gpu/lib/python3.9/site-packages/ncls/src/
> ncls.cpython-39-darwin.so, 0x0002): tried:
> '/Users/sriharshameghadri/miniforge3/envs/torch-gpu/lib/python3.9/site-packages/ncls/src/
> ncls.cpython-39-darwin.so' (mach-o file, but is an incompatible
> architecture (have 'x86_64', need 'arm64')),
> '/System/Volumes/Preboot/Cryptexes/OS/Users/sriharshameghadri/miniforge3/envs/torch-gpu/lib/python3.9/site-packages/ncls/src/
> ncls.cpython-39-darwin.so' (no such file),
> '/Users/sriharshameghadri/miniforge3/envs/torch-gpu/lib/python3.9/site-packages/ncls/src/
> ncls.cpython-39-darwin.so' (mach-o file, but is an incompatible
> architecture (have 'x86_64', need 'arm64'))
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pyranges/pyranges/discussions/346>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACTXRIFU5XNRGJGAY7E6Q7LXRJBMFANCNFSM6AAAAAA2SR5PXU>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Marco Mariotti
Ramón y Cajal Fellow, Group Leader
Department of Genetics, Microbiology and Statistics
Universitat de Barcelona
